### DISCUSSION TITLE
Persistence Events with async

### DISCUSSION THREAD

**Discussion by al-dpopowich** [sqlalchemy/sqlalchemy#7857] [Usage Questions] [answered]
Let's say I want to do some post-processing after a flush.  In classic, synchronous sqlalchemy I have often done something the following...

```python
from sqlalchemy import event
from sqlalchemy.orm import Session

@event.listens_for(Session, 'after_flush')
def receive_after_flush(session, flush_context):
   """For any mapper with `on_post_*()` defined, run the post-processing method"""
   # ON POST CREATE
   for obj in session.new:
      if hasattr(obj, 'on_post_create'):
         obj.on_post_create(session)

   # ON POST UPDATE
   for obj in session.dirty:
      if hasattr(obj, 'on_post_update'):
         obj.on_post_update(session)

   # ON POST DELETE
   for obj in session.deleted:
      if hasattr(obj, 'on_post_delete'):
         obj.on_post_delete(session)
```

My question: **_How do I do this with async_**?  I've seen the docs for the new `AdaptedConnection` and its `run_async` method, but I can't figure out how to do the same here.  Obviously, same goes for `before_flush`, etc.  I'm looking for an analogous `run_async` accessible from the AsyncSession which allows me to pass in an awaitable.

I am using postgresql w/ asyncpg.

Thanks!

**Comment by CaselIT** [comment, marked as answer]
Hi,

If I understand correctly you want `obj.on_post_delete` and similar to be async?

The session passed in the event is a normal synchronous one, so assuming you want to just query the db you can use synchronous style functions and they will be handled properly by sqlalchemy.

If you want to instead mix other types of io, like a network access using an async http library, I'm not sure if we have anything documented here. But you can get the adapted connection the documentation talks about using `session.connection().dbapi_connection`. Not very convenient but that's the current status.
Another alternative is to use the semi-private `await_only` function from `sqlalchemy.utils`

**Reply by zzzeek** [reply]
I've had other folks ask this question and Ive pointed them to await_only, so we will likely want to get a doc section for that also at some point.

**Reply by al-dpopowich** [reply]
Yes, while the above was just an example, my general question is how does one call async code _within_ event handlers which are sync.  `await_only` looks to be the answer!

Thank you!
